gf/include/alps/gf/mesh.hpp

```diff
@@ -779,6 +782,205 @@ namespace alps {
+
+            numerical_mesh(gf::statistics::statistics_type statistics=statistics::FERMIONIC):
+                    beta_(0.0), dim_(0), statistics_(statistics), valid_(false) {}
+
+            numerical_mesh(double b,  const std::vector<piecewise_polynomial<T,k> >&basis_functions,
+                           gf::statistics::statistics_type statistics=statistics::FERMIONIC):
+                    beta_(b), dim_(basis_functions.size()), basis_functions_(basis_functions), statistics_(statistics), valid_(false) {
+                set_validity();
+                //check_range();
+                compute_points();
+            }
+
+            int extent() const{
+                check_validity();
+                return dim_;
+            }
+
+            int operator()(index_type idx) const {
+                check_validity();
+                return idx();
+            }
+
+            /// Comparison operators
+            bool operator==(const numerical_mesh &mesh) const {
+                check_validity();
+                return beta_==mesh.beta_ && dim_==mesh.dim_ && statistics_==mesh.statistics_ && basis_functions_==mesh.basis_functions_;
+            }
+
+            /// Comparison operators
+            bool operator!=(const numerical_mesh &mesh) const {
+                check_validity();
+                return !(*this==mesh);
+            }
+
+            ///getter functions for member variables
+            double beta() const{ return beta_;}
+            statistics::statistics_type statistics() const{ return statistics_;}
+            const piecewise_polynomial<T,k>& basis_function(int l) const {
+                assert(l>=0 && l < dim_);
+                check_validity();
+                return basis_functions_[l];
+            }
+
+
+            /// Swaps this and another mesh
+            // It's a member function to avoid dealing with templated friend decalration.
+            void swap(numerical_mesh& other) {
+                using std::swap;
+                check_validity();
+                swap(this->beta_, other.beta_);
+                swap(this->dim_, other.dim_);
+                if (this->statistics_ != other.statistics_) {
+                    throw std::runtime_error("Do not swap numerical meshes with different statistics!");
+                }
+                swap(this->basis_functions_, other.basis_functions_);
+                base_mesh::swap(other);
+            }
+
+            void save(alps::hdf5::archive& ar, const std::string& path) const
+            {
+                check_validity();
+                ar[path+"/kind"] << "NUMERICAL_MESH";
+                ar[path+"/N"] << dim_;
+                ar[path+"/k"] << k;
+                if (boost::is_same<T,double>::value) {
+                    ar[path+"/scalar_type"] << "double";
+                } else if (boost::is_same<T,std::complex<double> >::value) {
+                    ar[path+"/scalar_type"] << "complex_double";
```

Comment:
If `T` is `std::complex<double>`, then the saved coefficients of the basis functions will anyway have "`__complex__`" attribute set, similar to the one described in https://github.com/ALPSCore/H5GF/blob/master/H5GF.rst#data-dataset . Do we need an extra "`scalar_type`" dataset here? 
Is there a situation when the `scalar_type` cannot be deduced from the type of `/basis_functions0/coefficients` dataset?
Or we want to increase data redundancy to make it more robust?